libs/shared-atomic-design-components/src/lib/atoms/CardTags.tsx

```diff
@@ -1,65 +1,114 @@
+  const hasAdditionalTags = (shortList || isMobile) && additionalTagsCount > 0
+
+  useEffect(() => {
+    const formattedItems = items.map(formatter)
+
+    const calculateChips = () => {
+      const parentWidth = parentRefDesktop.current?.offsetWidth
+      const gapSize = 16 // Adjust this based on the design (16px gap)
+      const sidePadding = 12 // Adjust this based on the design (12px padding each side )
+      const chipTotalMargin = 2 * sidePadding + gapSize
+
+      const chipWidths = formattedItems.map((chip) => {
+        const chipSpan = document.createElement('span')
+        chipSpan.style.visibility = 'hidden'
+        chipSpan.classList.add('chip')
+        chipSpan.innerHTML = chip
+
+        parentRefDesktop.current.appendChild(chipSpan)
+
+        const width = chipSpan.offsetWidth + chipTotalMargin
+
+        parentRefDesktop.current.removeChild(chipSpan)
+
+        return width
+      })
 
-  return (
-    <Tag.Group>
-      {tagList.map((tagId, i) => {
+      let totalWidth = 0
+      let visibleChipsCount = 0
+
+      for (let i = 0; i < chipWidths.length; i++) {
+        totalWidth += chipWidths[i]
+
+        if (totalWidth <= parentWidth) {
+          visibleChipsCount = i + 1
+        } else {
+          break
+        }
+      }
+
+      setVisibleChips(formattedItems.slice(0, visibleChipsCount))
+      setRemainingChips(formattedItems.slice(visibleChipsCount))
+    }
+
+    const throttledCalculateChips = debounce(calculateChips, 100)
+
+    throttledCalculateChips()
+
+    // Recalculate on window resize
+    window.addEventListener('resize', throttledCalculateChips)
+
+    return () => {
+      window.removeEventListener('resize', throttledCalculateChips)
+    }
+  }, [formatter, items])
+
+  return isMobile || shortList ? (
+    <div className="wrapper__mobile" ref={parentRefMobile}>
+      {itemsShowMobile.map((chip, i) => {
         const currentTag = (
-          <CardTag key={tagId}>{formatter ? formatter(tagId) : tagId}</CardTag>
+          <p key={chip} className="chip">
+            {chip}
+          </p>
         )
         const isLastVisibleTag = i === 2
 
         return hasAdditionalTags && isLastVisibleTag ? (
-          <div className="tags__last-row" key={tagId}>
+          <div className="wrapper__mobile last_row" key={chip}>
             {currentTag}
-            <CardTag key={`restNr-${i}`} className="tag--rest">
+            <p key={`restNr-${i}`} className="chip">
               {'+' + additionalTagsCount}
-            </CardTag>
+            </p>
           </div>
         ) : (
           currentTag
         )
       })}
-    </Tag.Group>
+    </div>
+  ) : (
+    <div className="wrapper__desktop" ref={parentRefDesktop}>
+      {visibleChips.map((chip) => (
+        <p key={chip} className="chip">
+          {chip}
+        </p>
+      ))}
+      {remainingChips.length > 0 && (
+        <p className="chip">+{remainingChips.length}</p>
+      )}
+    </div>
```

Comment:
We seem to have a bit too much happening here. I usually recommend not adding too much business logic in the definition of returned element in a component. If that's really necessary as seen here, then I'd recommend having smaller components with their returns cleaner.

Let me know if you'd like to pair on this, or try if you have some ideas already and I can give it another look.

Reply:
it sure looks busy here! I've divided them into 2 smaller components, in the last commit, let me know if it improves the final look :)